nucliadb_vectors/src/data_types/vector.rs

```diff
@@ -41,6 +41,54 @@ pub fn encode_vector(vec: &[Unit]) -> Vec<u8> {
         .fold(encode_length(vec![], vec), encode_unit)
 }
 
+#[inline]
+#[allow(unused)]
+pub fn simd_consine_similarity(mut x: &[u8], mut y: &[u8]) -> Dist {
+    use wide::f32x8;
+    const N: usize = 8;
+    let mut buff_x = [0; 8];
+    let mut buff_y = [0; 8];
+    x.read_exact(&mut buff_x).unwrap();
+    y.read_exact(&mut buff_y).unwrap();
+    let len_x = Len::from_le_bytes(buff_x);
+    let len_y = Len::from_le_bytes(buff_y);
+    assert_eq!(len_x, len_y);
+    let len = len_x;
+    let mut buff_x = [0; 4];
+    let mut buff_y = [0; 4];
+    let mut sum = 0.0;
+    let mut dem_x = 0.0;
+    let mut dem_y = 0.0;
+    let mut i = 0;
+    let mut x_simd_buffer = [0.0; N];
+    let mut y_simd_buffer = [0.0; N];
+    let mut free_in_buffer = 0;
+    while i < len {
+        i += 1;
+        x.read_exact(&mut buff_x).unwrap();
+        y.read_exact(&mut buff_y).unwrap();
+        x_simd_buffer[free_in_buffer] = Unit::from_le_bytes(buff_x);
+        y_simd_buffer[free_in_buffer] = Unit::from_le_bytes(buff_y);
+        free_in_buffer += 1;
+        if free_in_buffer == N || (i == len && free_in_buffer > 0) {
+            let partial_simd_x = f32x8::new(x_simd_buffer);
+            let partial_simd_y = f32x8::new(y_simd_buffer);
+            let partial_dot_product = partial_simd_x * partial_simd_y;
+            let partial_power_x = partial_simd_x * partial_simd_x;
+            let partial_power_y = partial_simd_y * partial_simd_y;
+            dem_x += partial_power_x.reduce_add();
```

Comment:
you don't want to reduce here. I think it can be done out of the loop.

Reply:
Yes I had a version with that out of the loop, but the performance stayed the same, I will do that version again and update the code!